src/gatsby/models/authors/parse-link.js

```diff
@@ -1,31 +1,77 @@
-const processStringLink = require('./string-link-processor.js')
+/*
+   Processor builder: makes a function that turns a string into processed Link metadata
+
+   Takes "Mod functions" as parameters, returns a function that starts with {} and pipes
+   that through each function, then returns the result.
+
+   signature: ({ scheme: string, host: string, pathname: string}, input)
+   example:      ^ https://      ^ mysite.com  ^ /rest/of/the/url ^ an object
+
+   The second parameter will be the current state of the input, and usually will
+   have to be spread into your return result.
+ */
+function processor(...mods) {
+  return function process(groups) {
+    return mods.reduce(
+      (currentResult, _, i) => mods[i](groups, currentResult),
+      {}
+    )
+  }
+}
+
+// Processor helpers
+
+function trimSlashes(input) {
+  return /^\/?(.*?)\/?$/.exec(input)[1]
+}
+
+const asSite = site => (groups, input) => ({
+  ...input,
+  site
+})
+
+const pathnameAsUsername = ({ pathname }, input) => ({
+  ...input,
+  username: trimSlashes(pathname)
+})
+
+const urlHTTPS = ({ host, pathname }, input) => ({
+  ...input,
+  url: 'https://' + host + pathname
+})
+
+const urlDefault = ({ scheme, host, pathname }, input) => ({
+  ...input,
+  url: (scheme || 'https://') + host + pathname
+})
 
 function WrongTypeError(type) {
   return new Error(`A ${type} cannot be used as input for parseLink!`)
 }
 
+// Building processors for recognized sites
+
+const processors = {
+  'twitter.com': processor(asSite('twitter'), urlHTTPS, pathnameAsUsername),
+  'github.com': processor(asSite('github'), urlHTTPS, pathnameAsUsername),
```

Comment:
I reminds me lodash flow (or may be something similar exists?)? should we just use that instead of writing our own?

Reply:
I actually started with lodash flow, but writing this with that ended up being more complicated specifically when sharing the parsed data. This implementation is a bit similar to Models, but sequential instead of parallel and return values matter.

Reply:
👍